pkg/signature/kms/cliplugin/common/interface.go

```diff
@@ -0,0 +1,81 @@
+//
+// Copyright 2021 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package common defines the JSON schema for plugin arguments and return values.
+package common
+
+import (
+	"crypto"
+	"time"
+)
+
+const (
+	// ProtocolVersion is the version of the schema and communication protocol for the plugin system.
+	// Breaking changes to the PluginClient and this schema necessarily mean major version bumps of
+	// this ProtocolVersion and the sigstore version.
+	// Plugin authors may choose to be backwards compatible with older versions.
+	ProtocolVersion            = "1"
+	DefaultAlgorithmMethodName = "defaultAlgorithm"
+	CreateKeyMethodName        = "createKey"
+	// TODO: Additonal methods to be implemented
+)
+
+// PluginArgs contains all the initialization and method arguments to be sent to the plugin as a CLI argument.
+type PluginArgs struct {
+	*MethodArgs
+	InitOptions *InitOptions `json:"initOptions"`
+}
+
+// InitOptions contains the initial arguments when calling cliplugin.LoadSignerVerifier().
+type InitOptions struct {
+	ProtocolVersion string      `json:"protocolVersion"`
+	KeyResourceID   string      `json:"keyResourceID"`
+	HashFunc        crypto.Hash `json:"hashFunc"`
+	// TODO: extracted values from signature.RPCOption from LoadSignerVerifier().
+}
+
+// MethodArgs contains the method arguments. MethodName must be specified,
+// while any one of the other fields describing method arguments must also be specified.
+type MethodArgs struct {
+	// MethodName specifies which method is intended to be called.
+	MethodName       string                `json:"methodName"`
+	DefaultAlgorithm *DefaultAlgorithmArgs `json:"defaultAlgorithm,omitempty"`
```

Comment:
Given this returns a constant string for a given plugin, can we switch this to a field in InitOptions, and then `DefaultAlgorithm()` should just return that string?

Reply:
We could, but it still requires invoking the plugin first. And it seems like `DefaultAlgorithm()` is used by cosign only when calling `CreateKey()`.

Reply:
You're right, I was thinking this was something that could be statically defined across plugins but each plugin has their own default algorithm.

I think there are some improvements we can make outside of this PR to change the behavior of `CreateKey` to use a default if no algorithm is specified, but given the other thread about deprecating it entirely, we can remove `DefaultAlgorithm` at the same time.

Reply:
We could consider removing the method from the plugin as a breaking change. There may be people that use it, but we don’t have the visibility on their use cases.

Reply:
I think we should document what is required (namely, SignMessage), and what is optional, so plugin authors know where to focus their development.

We can circle back to this before we cut a release with this plugin, and see about removing functionality from Cosign.